### ISSUE TITLE
problem synching state when using async onJoin

### ISSUE THREAD

**Issue by endel** [colyseus/colyseus#137]
The scenario I'm having is:

- `onInit()` is async. The state is set only after a timeout.
- `onJoin()` is async. The state is mutated once it's ready.

The problem:

- The first client [doesn't receive the first state](https://github.com/gamestdio/colyseus/blob/f3f0c217eed04a0c1a44821a833954675b85971f/src/Room.ts#L359-L362), and once it receives the first patch, an error occurs, since there's no first state to apply the patch.

```
fossil-delta.js:399 Uncaught TypeError: Cannot read property 'length' of undefined
    at Object.fossilDelta.apply
```

The workaround I've come up with is manually sending the first state when it's ready:

```typescript
onInit () {
  this.whenStateIsReady = new Promise((resolve, reject) => {
    setTimeout(() => {
      this.setState({ foo: "bar" });
      resolve();
    }, 1000);
  });
}

async onJoin(client) {
  await this.whenStateIsReady;

  // this is required to prevent client-side error
  this.sendState(client);
}
```

Would be great if the framework can handle this situation internally.

**Comment by Sahbum** [comment]
I was having this issue last night - the onInit function was loading stuff from a database, but the onJoin was finishing before onInit was done, so the first client was getting an inconsistent state.  I'll have to use the above to solve my issue.

It would be great if the other events for a room were held until onInit was completed.

**Comment by gioragutt** [comment]
This would require [this code](https://github.com/gamestdio/colyseus/blob/df63b22ad3fabb313478e0c550384ab59c9316fd/src/MatchMaker.ts#L257) to handle it being a promise, which would require the entire wrapping method to be `async`, which breaks API,

If they fix it, it will probably be for `v1.0`, but not before.
This is probably a right thing to do, usually you want to allow async code.
